src/main/java/graphql/execution/ValuesResolverConversion.java

```diff
@@ -69,11 +69,17 @@ static Object valueToLiteralImpl(GraphqlFieldVisibility fieldVisibility,
             if (valueMode == NORMALIZED) {
                 return assertShouldNeverHappen("can't infer normalized structure");
             }
-            return ValuesResolverLegacy.valueToLiteralLegacy(
+            Value<?> value = ValuesResolverLegacy.valueToLiteralLegacy(
                     inputValueWithState.getValue(),
                     type,
                     graphqlContext,
                     locale);
+            //
+            // the valueToLiteralLegacy() nominally cant know if null means never set or is set to a null value
+            // but this code can know - its is SET to a value so, it MUST be a Null Literal
+            // this method would assert at the end of it if inputValueWithState.isNotSet() were true
+            //
+            return value == null ? NullValue.of() : value;
```

Comment:
The comment outlines why this is safe - but what you can see in this diff is `if (inputValueWithState.isInternal())` being called.  Since this is true then the value MUST have been set into the input value and hence we can turn the legacy null into a AST NullValue safely.